src/modules/sync/lib/convert-tree-to-sync-options.ts

```diff
@@ -12,19 +13,124 @@ type PushOptionsWithSelections = {
+		if ( normalizedPath === exclusion || normalizedPath.startsWith( exclusion ) ) {
+			return true;
+		}
+	}
+
+	return false;
 };
 
-const collectPathIds = ( nodes: TreeNode[], pathIds: string[] = [] ): string[] => {
+const getPathSyncOption = ( path: string ): SyncOption | null => {
+	if ( shouldExcludePathFromSync( path ) ) {
+		return null;
+	}
+
+	const normalizedPath = normalizePath( path );
+
+	for ( const { pattern, option } of PATH_SYNC_MAP ) {
+		if ( pattern.test( normalizedPath ) ) {
+			return option;
+		}
+	}
+
+	if ( normalizedPath ) {
+		return SYNC_OPTIONS.contents;
+	}
+
+	return null;
+};
+
+const collectNodes = < T >( nodes: TreeNode[], extractor: ( node: TreeNode ) => T | null ): T[] => {
+	const collector: T[] = [];
 	nodes.forEach( ( node ) => {
-		if ( node.checked && node.pathId ) {
-			pathIds.push( node.pathId );
+		if ( node.checked ) {
+			const value = extractor( node );
+			if ( value ) collector.push( value );
 		} else if ( node.indeterminate && node.children ) {
-			collectPathIds( node.children, pathIds );
+			collector.push( ...collectNodes( node.children, extractor ) );
 		}
 	} );
-	return pathIds;
+	return collector;
+};
+
+const collectPathIds = ( nodes: TreeNode[] ): string[] =>
+	collectNodes( nodes, ( node ) => node.pathId || null );
+
+const collectSelectedPaths = ( nodes: TreeNode[] ): string[] =>
+	collectNodes( nodes, ( node ) =>
+		node.path && ! shouldExcludePathFromSync( node.path ) ? node.path : null
+	);
+
+const groupPathsBySyncOption = ( paths: string[] ): Record< SyncOption, boolean > => {
+	const syncOptions: Record< SyncOption, boolean > = {
+		[ SYNC_OPTIONS.all ]: false,
+		[ SYNC_OPTIONS.sqls ]: false,
+		[ SYNC_OPTIONS.paths ]: false,
+		[ SYNC_OPTIONS.themes ]: false,
+		[ SYNC_OPTIONS.plugins ]: false,
+		[ SYNC_OPTIONS.uploads ]: false,
+		[ SYNC_OPTIONS.contents ]: false,
+	};
+
+	paths.forEach( ( path ) => {
+		const syncOption = getPathSyncOption( path );
+		if ( syncOption ) {
+			syncOptions[ syncOption ] = true;
+		}
+	} );
+
+	return syncOptions;
+};
+
+const extractSpecificSelections = (
+	paths: string[]
+): PushOptionsWithSelections[ 'specificSelections' ] => {
+	const specificSelections: PushOptionsWithSelections[ 'specificSelections' ] = {};
+	const collections = {
+		plugins: new Set< string >(),
+		themes: new Set< string >(),
+		uploads: new Set< string >(),
+	};
+
+	paths.forEach( ( path ) => {
+		const normalizedPath = normalizePath( path );
+
+		Object.entries( SPECIFIC_PATTERNS ).forEach( ( [ key, pattern ] ) => {
+			const match = normalizedPath.match( pattern );
+			if ( match ) {
+				collections[ key as keyof typeof collections ].add( match[ 1 ] );
+			}
+		} );
+	} );
+
+	Object.entries( collections ).forEach( ( [ key, set ] ) => {
+		if ( set.size > 0 ) {
+			specificSelections[ key as keyof typeof collections ] = Array.from( set );
+		}
+	} );
+
+	return Object.keys( specificSelections ).length > 0 ? specificSelections : undefined;
 };
```

Comment:
If I select a file like `wp-content/example.jpg` in my sync push, it won't currently be included in the push archive. We should either change this behavior so that it's sent to the server, or update the file modal to not display files if they don't live inside the `plugins`, `themes` or `uploads` folders. 

EDIT: I see that @sejas also commented on this https://github.com/Automattic/studio/pull/1940#pullrequestreview-3403443147